### PR TITLE
Kubernetes Pod support

### PR DESCRIPTION
**Pull request by rpitonak**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -40,7 +41,8 @@ test: build-test-container test-in-container test-doc-examples
 
 test-in-container:
 	@# use it like this: `make test-in-container TEST_TARGET=tests/integration/test_utils.py`
-	docker run --net=host --rm -v /dev:/dev:ro -v /var/lib/docker:/var/lib/docker:ro --security-opt label=disable --cap-add SYS_ADMIN -ti -v /var/run/docker.sock:/var/run/docker.sock -v $(CURDIR):/src -v $(CURDIR)/pytest-container.ini:/src/pytest.ini $(TEST_IMAGE_NAME) make exec-test TEST_TARGET=$(TEST_TARGET)
+	sed -e s#"${HOME}"#/root#g ${HOME}/.kube/config > $(kubedir)/config
```
Comment (TomasTomecek) [inline Makefile]: oh boy, do you really want to change files in my $HOME?

File: conu/apidefs/image.py [hunk 2]
```diff
@@ -170,6 +175,53 @@ def create_container(self, container_params):
...
+        if image_data.exposed_ports is not None:
+            for port in image_data.exposed_ports:
+                try:
+                    protocol = port.split("/", 1)[1]
+                    exposed_ports.append(client.V1ContainerPort(container_port=port.split("/", 1)[0],
+                                                                protocol=protocol))
+                except IndexError:  # protocol is not defined in image metadata
+                    exposed_ports.append(client.V1ContainerPort(container_port=port.split("/", 1)[0]))
+
+        # generate random container name
```
Comment (TomasTomecek) [inline conu/apidefs/image.py]: would it be possible to make it not random? how about something like '{image-name}-{username}-{random-4-letters}'

File: Makefile [hunk 3]
```diff
@@ -40,7 +41,8 @@ test: build-test-container test-in-container test-doc-examples
...
+	docker run --net=host --rm -v /dev:/dev:ro -v /var/lib/docker:/var/lib/docker:ro --security-opt label=disable --cap-add SYS_ADMIN -ti -v /var/run/docker.sock:/var/run/docker.sock -v $(CURDIR):/src -v $(CURDIR)/pytest-container.ini:/src/pytest.ini -v ${HOME}/.minikube:/root/.minikube -v $(kubedir):/root/.kube $(TEST_IMAGE_NAME) make exec-test TEST_TARGET=$(TEST_TARGET)
```
Comment (TomasTomecek) [inline Makefile]: some docs on how to set up minikube would be nice

File: Makefile (hunk 1 above)
Comment (dhodovsk) [reply Makefile]: @TomasTomecek could this affect the source file (the one in home)? why is this not safe?

Comment (TomasTomecek) [reply Makefile]: Oh, my bad! I thought it's `-i $HOME...`. Should work just fine.

File: conu/apidefs/image.py [hunk 4]
```diff
@@ -170,6 +175,53 @@ def create_container(self, container_params):
...
+                                       env=env_variables,
+                                       image=image_data.name,
+                                       name=container_name,
+                                       ports=exposed_ports)
+
+        pod_metadata = client.V1ObjectMeta(name=container_name + "-pod")
+        pod_spec = client.V1PodSpec(containers=[container])
+        pod = client.V1Pod(spec=pod_spec, metadata=pod_metadata)
+
+        pod_instance = api.create_namespaced_pod(namespace=namespace, body=pod)
```
Comment (dhodovsk) [inline conu/apidefs/image.py]: There is an error when using container with exposed ports, e.g. `docker.io/nginx`

Log:
```
kubernetes.client.rest.ApiException: (400)
Reason: Bad Request
HTTP response headers: HTTPHeaderDict({'Content-Type': 'application/json', 'Date': 'Wed, 30 May 2018 12:19:27 GMT', 'Content-Length': '510'})
HTTP response body: {
"kind":"Status",
"apiVersion":"v1",
"metadata":{},
"status":"Failure",
"message":"Pod in version \"v1\" cannot be handled as a Pod: v1.Pod: Spec:
v1.PodSpec: Containers: []v1.Container: v1.Container: Ports:
[]v1.ContainerPort: v1.ContainerPort: ContainerPort:
readUint32: unexpected character: \ufffd, error found in #10 byte of ...|erPort\": \"80\",
\"prot|..., bigger context ...|name\": \"figpwbxmbt\", \"ports\":
[{\"containerPort\": \"80\", \"protocol\": \"tcp\"}]}]}}|...",
"reason":"BadRequest","code":400}

```

**Comment by dhodovsk** [comment]
Please also add some logging for better troubleshooting.

File: conu/apidefs/image.py [hunk 5]
```diff
@@ -170,6 +180,61 @@ def create_container(self, container_params):
         """
         raise NotImplementedError("create_container method is not implemented")
 
+    def run_in_pod(self, namespace="default"):
```
Comment (TomasTomecek) [inline conu/apidefs/image.py]: for now, we should probably put this method on DockerImage

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: failure, return code = __1__ 
* Result url: __[https://da.gd/7HYyd](https://da.gd/7HYyd)__
* Comment: __See details for logs__

File: conu/backend/k8s/pod.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        delete pod from the Kubernetes cluster
+        :return:
+        """
+        body = client.V1DeleteOptions()
+
+        try:
+            status = api.delete_namespaced_pod(self.name, self.namespace, body)
+            if status.status == 'Failure':
+                ConuException("Deletion of Pod failed")
```
Comment (jpopelka) [inline conu/backend/k8s/pod.py]: Does it really work without `raise` ?

File: conu/backend/k8s/pod.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+            logger.info("Deleting Pod {pod_name} in {namespace}".format(pod_name=self.name,
+                                                                        namespace=self.namespace))
+
+            self.phase = PodPhase.TERMINATING
+
+        except ApiException as e:
+            ConuException("Exception when calling Kubernetes API - delete_namespaced_pod: {}\n".format(e))
```
Comment (jpopelka) [inline conu/backend/k8s/pod.py]: The `try` section is too broad. The following should be enough.

```python
try:
  status = api.delete_namespaced_pod(self.name, self.namespace, body)
except ApiException as e:
  raise ConuException('...')

if status.status == 'Failure':
  raise ConuException("Deletion of Pod failed")
...
```

File: conu/backend/k8s/pod.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+    def get_status(self):
+        """
+        get status of the Pod
+        :return: V1PodStatus, https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1PodStatus.md
+        """
+        try:
+            api_response = api.read_namespaced_pod_status(self.name, self.namespace)
+            return api_response.status
```
Comment (jpopelka) [inline conu/backend/k8s/pod.py]: The same as above, i.e. I'd move the `return` outside the `try:` section.

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: failure, return code = __1__ 
* Result url: __[https://da.gd/IQ2a](https://da.gd/IQ2a)__
* Comment: __See details for logs__

**Comment by dhodovsk** [comment]
We should make the CI test results green. Now it fails on missing config file, since kubernetes/minikube was not installed. We could:

1. Add new target to Makefile - `make install-minikube`
2, Make minikube installation only as part of CI
3, provide custom config file with pointers to own kubernetes instance, where every test would have own namespace.

**Comment by TomasTomecek** [comment]
Sounds good.

File: requirements.sh [hunk 9]
```diff
@@ -15,3 +15,14 @@ dnf install -y acl docker libselinux-utils \
...
 # pyxattr has different naming in Fedora 27 and Fedora 28
 dnf install -y python2-pyxattr || dnf install -y pyxattr
+
+# https://github.com/kubernetes/minikube
+SKIP_INSTALL_MINIKUBE=${SKIP_INSTALL_MINIKUBE:-false}
+
+if [ "$SKIP_INSTALL_MINIKUBE" = false ] ; then
+    dnf install libvirt-daemon-kvm qemu-kvm
+    dnf install kubernetes
+    curl -LO https://storage.googleapis.com/minikube/releases/latest/docker-machine-driver-kvm2 && chmod +x docker-machine-driver-kvm2 && mv docker-machine-driver-kvm2 /usr/local/bin/
```
Comment (TomasTomecek) [inline requirements.sh]: since we have `set -e` in this script, you can put each command on a new line

Comment (TomasTomecek) [inline requirements.sh]: blah, they still use docker-machine

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: failure, return code = __1__ 
* Result url: __[https://da.gd/nfNV](https://da.gd/nfNV)__
* Comment: __See details for logs__

File: Makefile [hunk 10]
```diff
@@ -41,8 +41,7 @@ test: build-test-container test-in-container test-doc-examples
 
 test-in-container:
 	@# use it like this: `make test-in-container TEST_TARGET=tests/integration/test_utils.py`
-	sed -e s#"${HOME}"#/root#g ${HOME}/.kube/config > $(kubedir)/config
-	docker run --net=host --rm -v /dev:/dev:ro -v /var/lib/docker:/var/lib/docker:ro --security-opt label=disable --cap-add SYS_ADMIN -ti -v /var/run/docker.sock:/var/run/docker.sock -v $(CURDIR):/src -v $(CURDIR)/pytest-container.ini:/src/pytest.ini -v ${HOME}/.minikube:/root/.minikube -v $(kubedir):/root/.kube $(TEST_IMAGE_NAME) make exec-test TEST_TARGET=$(TEST_TARGET)
+	docker run --net=host --rm -v /dev:/dev:ro -v /var/lib/docker:/var/lib/docker:ro --security-opt label=disable --cap-add SYS_ADMIN -ti -v /var/run/docker.sock:/var/run/docker.sock -v $(CURDIR):/src -v $(CURDIR)/pytest-container.ini:/src/pytest.ini $(TEST_IMAGE_NAME) oc login 127.0.0.1:8443 -u developer -p developer -n conu --insecure-skip-tls-verify && make exec-test TEST_TARGET=$(TEST_TARGET)
```
Comment (jpopelka) [inline Makefile]: I guess you want ([see also](https://stackoverflow.com/questions/28490874/docker-run-image-multiple-commands))
`docker run ... $(TEST_IMAGE_NAME) /bin/bash -c "oc login ... && make exec-test ..."`

File: tests/integration/test_k8s.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+from ..constants import FEDORA_MINIMAL_REPOSITORY, FEDORA_MINIMAL_REPOSITORY_TAG
+
+
+def test_pod():
+    with DockerBackend() as backend:
+        image = backend.ImageClass(FEDORA_MINIMAL_REPOSITORY, tag=FEDORA_MINIMAL_REPOSITORY_TAG)
+
+        pod = image.run_in_pod(namespace='conu')
+
+        pod.wait(200)
```
Comment (dhodovsk) [inline tests/integration/test_k8s.py]: This should go also to try block.

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: failure, return code = __1__ 
* Result url: __[https://da.gd/91z8](https://da.gd/91z8)__
* Comment: __See details for logs__

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: failure, return code = __1__ 
* Result url: __[https://da.gd/j07l](https://da.gd/j07l)__
* Comment: __See details for logs__

**Comment by dhodovsk** [comment]
Tests are passing locally. Can we merge, @TomasTomecek @jpopelka ?

File: README.md [hunk 12]
```diff
@@ -105,6 +105,9 @@ modularitycontainers/conu:0.2.0 python3 /app/my_source.py
     - delay execution
     - delay between iterations
 
+# Kubernetes
+
+Running tests in Kubernetes requires Minikube installed on a host machine. See installation guide in `README.md` of the [minikube Github repository](https://github.com/kubernetes/minikube).
```
Comment (jpopelka) [inline README.md]: We don't use Minikube any more, do we ?

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: failure, return code = __1__ 
* Result url: __[https://da.gd/qOg0](https://da.gd/qOg0)__
* Comment: __See details for logs__

File: Makefile [hunk 13]
```diff
@@ -4,6 +4,7 @@ CONU_REPOSITORY := docker.io/usercont/conu:dev
 TEST_IMAGE_NAME := conu-tests
 DOC_EXAMPLE_PATH := "docs/source/examples"
 VERSION := 0.4.0
+kubedir := $(shell mktemp -d /tmp/tmp.conu-kube-XXXXX)
```
Comment (TomasTomecek) [inline Makefile]: doesn't seem to be used anywhere; let's not create tons of useless dirs

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: failure, return code = __1__ 
* Result url: __[https://da.gd/T5fv2](https://da.gd/T5fv2)__
* Comment: __See details for logs__

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: failure, return code = __1__ 
* Result url: __[https://da.gd/osUcc](https://da.gd/osUcc)__
* Comment: __See details for logs__

**Pull request merged**